src/async_wrap.cc

```diff
@@ -452,6 +452,15 @@ static void EnablePromiseHook(const FunctionCallbackInfo<Value>& args) {
   }
 }
 
+static void SetPromiseHooks(const FunctionCallbackInfo<Value>& args) {
+  Environment* env = Environment::GetCurrent(args);
+  Local<Context> ctx = env->context();
+  Local<Value> init_hook = args[0];
+  Local<Value> before_hook = args[1];
+  Local<Value> after_hook = args[2];
+  Local<Value> resolve_hook = args[3];
+  ctx->SetPromiseHooks(init_hook, before_hook, after_hook, resolve_hook);
+}
 
 static void DisablePromiseHook(const FunctionCallbackInfo<Value>& args) {
```

Comment:
`EnablePromiseHook()`/`DisablePromiseHook()` still suffer from the problem mentioned in the design doc that they perform per-Isolate actions, when what we want are per-Context actions… we could make that go away with the new hook functions, though, right? We’d just need to make the `init_hook` function here do a call into C++ to set up the GC tracking if there any `destroy` hooks, correct?

Reply:
Yes, the destroy-hook fallback remains unchanged currently, so it still suffers from the isolate issues. I'm hoping to eliminate that eventually, but it may have to wait until later. My plan is to eliminate the switch to native hooks for destroy hooks and instead just apply gc-tracking from the JS side. Might make the gc-tracking a bit more expensive, but I suspect the benefit of the _rest_ of the hooks moving the JS should more than offset that.